daml-assistant/src/DA/Daml/Assistant/Cache.hs

```diff
@@ -63,21 +63,25 @@ cacheAvailableSdkVersions
     -> IO [SdkVersion]
     -> IO ([SdkVersion], CacheAge)
 cacheAvailableSdkVersions DontUseCache getVersions = (, Fresh) <$> getVersions
-cacheAvailableSdkVersions UseCache { forceReload, cachePath, damlPath } getVersions = do
+cacheAvailableSdkVersions UseCache { overrideTimeout, cachePath, damlPath } getVersions = do
     damlConfigE <- tryConfig $ readDamlConfig damlPath
     let configUpdateCheckM = join $ eitherToMaybe (queryDamlConfig ["update-check"] =<< damlConfigE)
-        updateCheck | forceReload = UpdateCheckEvery (CacheTimeout 1)
-                    | Just updateCheck <- configUpdateCheckM = updateCheck
-                    | otherwise = UpdateCheckEvery (CacheTimeout 86400) -- One day, in seconds
-    case updateCheck of
-        UpdateCheckNever -> do
-            valueAgeM <- loadFromCacheWith cachePath versionsKey (CacheTimeout 0) deserializeVersions
-            pure $ fromMaybe ([], Stale) valueAgeM
-
-        UpdateCheckEvery timeout ->
-            cacheWith cachePath versionsKey timeout
-                serializeVersions deserializeVersions
-                getVersions
+        mbTimeout
+            -- A few different things can override the timeout behaviour in
+            -- daml-config.yaml, chiefly `daml install latest` and `daml version --force-reload yes`
+          | Just timeout <- overrideTimeout = Just timeout
+          | Just updateCheck <- configUpdateCheckM =
+              case updateCheck of
+                -- When UpdateCheckNever, still refresh the cache if it doesn't exist
+                UpdateCheckNever -> Nothing
+                UpdateCheckEvery timeout -> Just timeout
+          | otherwise = Just (CacheTimeout 86400)
+    case mbTimeout of
+      Nothing -> pure ([], Stale)
+      Just timeout ->
+        cacheWith cachePath versionsKey timeout
+            serializeVersions deserializeVersions
+            getVersions
```

Comment:
This function doesn't call loadFromCache anymore - is that a problem?

Reply:
I was also surprised by this but then saw that `cacheWith` does call `loadFromCache` as its first action, so I think this is actually cleaner

Reply:
Yeah - I decided to move the logic into cacheWith - this code in general had a few spots where logic used by a positive branch was being reimplemented by a negative branch.